### PR TITLE
Add support for Github Enterprise

### PR DESCRIPTION
**Pull request by Line-Noise**
This PR adds basic support for Github Enterprise through a `GITHUB_URL` environment variable.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,6 +21,9 @@ Alternately, you can download via `go get github.com/clever/microplane/cmd`. In
 
 The `GITHUB_API_TOKEN` environment variable must be set for Github. This should be a [GitHub Token](https://github.com/settings/tokens) with `repo` scope.
 
+Optionally: The `GITHUB_URL` environment variable can be set to use a Github Enterprise setup, otherwise it will use https://github.com.
+The `GITHUB_URL` **must** be a valid URL for the API endpoint including a trailing slash: e.g. `https://git.yourcompany.com/api/v3/`
```
Comment (nathanleiby) [inline README.md]: nit: consistency re: backticks + URLs ... suggest using backticks around `https://github.com` too.

File: initialize/initialize.go [hunk 2]
```diff
@@ -7,7 +7,8 @@ import (
 	"log"
 	"os"
 	"sort"
-	"strings"
+  "strings"
+	"net/url"
```
Comment (nathanleiby) [inline initialize/initialize.go]: needs a `go fmt`

File: initialize/initialize.go [hunk 3]
```diff
@@ -128,6 +129,13 @@ func githubSearch(query string) ([]Repo, error) {
 
 	client := github.NewClient(tc)
 
+	if os.Getenv("GITHUB_URL") != "" {
+		baseEndpoint, _ := url.Parse(os.Getenv("GITHUB_URL"))
```
Comment (nathanleiby) [inline initialize/initialize.go]: we should return + handle these errors

File: push/push.go [hunk 4]
```diff
@@ -101,6 +101,13 @@ func GithubPush(ctx context.Context, input Input, repoLimiter *time.Ticker, push
 	tc := oauth2.NewClient(ctx, ts)
 	client := github.NewClient(tc)
 
+	if os.Getenv("GITHUB_URL") != "" {
+		baseEndpoint, _ := url.Parse(os.Getenv("GITHUB_URL"))
+		client.BaseURL = baseEndpoint
+		uploadEndpoint, _ := url.Parse(os.Getenv("GITHUB_URL") + "upload/")
+		client.UploadURL = uploadEndpoint
+	}
```
Comment (nathanleiby) [inline push/push.go]: by "rule of 3" would love if we could refactor this logic (perhaps all of `createGithubClient`) into a helper fn so it's not repeated throughout

File: initialize/initialize.go [hunk 5]
```diff
@@ -155,12 +163,18 @@ func githubSearch(query string) ([]Repo, error) {
...
+	}
+
 	repos := []Repo{}
 	for _, r := range allRepos {
 		repos = append(repos, Repo{
 			Name:     r.GetName(),
 			Owner:    r.Owner.GetLogin(),
-			CloneURL: fmt.Sprintf("git@github.com:%s", r.GetFullName()),
+			CloneURL: fmt.Sprintf("git@%s:%s", hostname, r.GetFullName()),
 			Provider: "github",
```
Comment (nathanleiby) [inline initialize/initialize.go]: Ultimately, I think I would like to refactor this a bit.. No need to do that within this PR but what do you think of the following? 

I believe a given microplane flow is with only one provider. If so, then we can abstract out the idea of a provider a bit more, make it more explicit, and fix edge cases (e.g. currently, it's possible that the env var value could change and the API URL would be different between `push` and `merge`!)

In the future, a user might run: `mp init --provider github --provider-url <your API URL>` (these flags would be optional and have defaults) 

Then, we can store the provider and providerURL on `initialize.Output`
https://github.com/Clever/microplane/blob/f9bb2bd50cb558749c4023c518bea701747cf561/initialize/initialize.go#L34-L39

When doing all future steps, we use the stored provider values. e.g. we can make `CloneURL()` a function  that generates a string based on provider values, and we can remove URL parsing logic in later steps.

File: initialize/initialize.go (hunk 3 above)
Comment (Line-Noise) [reply initialize/initialize.go]: I basically cribbed the code direct from https://github.com/google/go-github/blob/master/github/github.go#L310

Perhaps we should be using `NewEnterpriseClient` if `GITHUB_URL` is defined and let it do the error handling?

File: push/push.go (hunk 4 above)
Comment (Line-Noise) [reply push/push.go]: Yes, I was thinking this as I was adding the same code in all these different places.

File: initialize/initialize.go (hunk 5 above)
Comment (Line-Noise) [reply initialize/initialize.go]: This is a great idea! I didn't think about what would happen if someone changed `GITHUB_URL` part way through the process. It should definitely be made idempotent.

File: initialize/initialize.go (hunk 2 above)
Comment (Line-Noise) [reply initialize/initialize.go]: Oops. That'll teach me to edit stuff directly in GitHub's editor. It converted the tab to spaces w
en I fixed the merge conflict. I'll fix it up.

**Comment by Line-Noise** [comment]
I'll fix up a few things based on your comments but I'm not a `go` programmer (I'm not a programmer at all, I'm a sysadmin!). This is my first time touching `go`. :-)

File: initialize/initialize.go (hunk 3 above)
Comment (nathanleiby) [reply initialize/initialize.go]: I like the idea of using `NewEnterpriseClient` 👍

**Comment by nathanleiby** [comment]
Pretty impressive that you jumped right into a new language!

**Comment by VasuBhaskar** [comment]
@nathanleiby it would be great if the support for git enterprise can be merged. This tool can be used widely only if supports enterprise

**Comment by VasuBhaskar** [comment]
I started to use this and eventually realized, that there is no support for enterprise :-(

**Comment by gtorre** [comment]
@nathanleiby @Line-Noise any chance this work will be picked up again?

**Comment by nathanleiby** [comment]
@gtorre Thanks for the follow-up! When I took a look awhile back, I wasn't able to make progress because I didn't have a Github enterprise account to test against. 

It looks like I'm able to make a 14d free trial for Github Enterprise, so I'll try that and get back to you if I'm able to make this work.

**Comment by gtorre** [comment]
Thanks @nathanleiby !

**Comment by nathanleiby** [comment]
@Line-Noise thanks for starting this so long ago and apologies we weren't able to make it happen at the time. 

I'm revisiting this now and have created two new PRs

1. refactoring the provider: https://github.com/Clever/microplane/pull/73
2. then adding support for Github enterprise: https://github.com/Clever/microplane/pull/74

I'll close this PR for now, given that the same work lives on in those. Thank you again!

cc @gtorre @VasuBhaskar

**Pull request closed without merging**